qpmodel/optimizer.cs

```diff
@@ -342,102 +425,135 @@ public void ExploreGroup(Memo memo)
             explored_ = true;
         }
 
-        // propagate the property requirement from the top
-        public void PropagateProperty(PhysicProperty property)
+        // preprocess exprList_ to get whether property can be supplied and subproperties
+        // also find members that can propagate required property and split them
+        internal void GnenerateCandidates(PhysicProperty property, out List<CGroupMember> candidates,
+            out List<bool> fulfilled, out List<List<PhysicProperty>> subproplist)
         {
-            return;
-        }
+            candidates = new List<CGroupMember>();
+            fulfilled = new List<bool>();
+            subproplist = new List<List<PhysicProperty>>();
 
-        // scan through the member list and return the least cost member
-        public CGroupMember locateMinCostMember()
-        {
-            double mincost = double.MaxValue;
-            foreach (var v in exprList_)
+            foreach (var member in exprList_)
             {
-                var physic = v.physic_;
-                if (physic != null && physic.Cost() < mincost)
+                if (member.physic_ is null) continue;
+
+                var subproperties = ProcessChildProperty(member.physic_, property, out bool propagated);
+                if (propagated)
                 {
-                    mincost = physic.Cost();
-                    minMember_ = v;
+                    candidates.Add(member);
+                    fulfilled.Add(true);
+                    subproplist.Add(subproperties);
+                    candidates.Add(member);
+                    fulfilled.Add(false);
+                    subproplist.Add(new List<PhysicProperty> (new PhysicProperty[member.physic_.children_.Count]));
+                }
+                else
+                {
+                    candidates.Add(member);
+                    fulfilled.Add(property?.IsPropertySupplied(member.physic_) ?? true);
+                    subproplist.Add(subproperties);
                 }
             }
-
-            minIncCost_ = mincost;
-            Debug.Assert(IsSolverOptimizedGroup() ||
-                minMember_.physic_.InclusiveCost() == minIncCost_);
-            return minMember_;
+            Debug.Assert(candidates.Count == fulfilled.Count && candidates.Count == subproplist.Count);
```

Comment:
add assertion here:
if property is null, candidate number = physic member number